### PR TITLE
OdbcHook returns None. Related to #15016 issue.

### LINKED ISSUE #15016: OdbcHook string values in connect_kwargs dict converts to None
**Apache Airflow version**: 2.0.1

**What happened**:

`OdbcHook` returns `None` for non-boolean-like string values in `connect_kwargs` dict arg

**What you expected to happen**:

`connect_kwarg` values should remain as is.

**How to reproduce it**:

```python
>>> from airflow.providers.odbc.hooks.odbc import OdbcHook
>>> OdbcHook('my_conn', connect_kwargs={'CurrentSchema': 'SCHEMA'}).connect_kwargs
{'CurrentSchema': None}
```

**Anything else we need to know**:

The issue lies in: https://github.com/apache/airflow/blob/db9febdb3be97832679d2ced8028fd7f1c21cd4e/airflow/providers/odbc/hooks/odbc.py#L170-L173

There's no `else` block that returns `val`. As it is, any string value will instead return `None`.

Pardon my ignorance on the subject, but this raises a question: Why is `clean_bool` being called in the first place for a user-provided dictionary? I'm not sure how this is necessary because the user can provide a literal boolean value in the dictionary if needed, no? If in the event that a driver needs to take a case-sensitive boolean string for some parameter, then `clean_bool` would make it impossible to provide such a value.

### PR DESCRIPTION
**Pull request by Goodkat**
This PR is related to #15016 issue.
OdbcHook returns None for non-boolean-like string values in connect_kwargs dict arg, however connect_kwarg values should remain as is in this case.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/providers/odbc/hooks/odbc.py
Comment (turbaszek) [inline airflow/providers/odbc/hooks/odbc.py]: With this change we may be able to remove the pylint comment from L168, WDYT?

**Comment by marcosmarxm** [comment]
Nice @Goodkat! Can you add a simple test for this? https://github.com/apache/airflow/blob/master/tests/providers/odbc/hooks/test_odbc.py already has tests for other connections functionalities and add this will helpfully

File: airflow/providers/odbc/hooks/odbc.py
Comment (Goodkat) [reply airflow/providers/odbc/hooks/odbc.py]: There are different types (at least boolean and string) that could be returned by this function. Therefore (I suppose) the code analyser complains of this fact.
However, as I realized, it could be acceptable sometimes in dynamic languages, so we can conditionally return different data types.
In our case the clean_bool() function's result is used as a value in the dictionary and dictionaries itself can contain the keys/values of the different types. Therefore we may ignore this warning and remove the pylint comment from L168.
Please tell me if it does make sense for you.

**Comment by marcosmarxm** [comment]
@Goodkat can you check discussion on #15016? Maybe is possible to remove this function and change documentation to use JSON object with true/false values. If is possible removing the function will turn the code cleaner =)

File: airflow/providers/odbc/hooks/odbc.py
Comment (Goodkat) [reply airflow/providers/odbc/hooks/odbc.py]: As the clean_bool is planned to be removed, the pylint comment will be cleaned as well.

**Comment by eladkal** [comment]
@Goodkat can you fix the failing test?

**Review by dstandish — COMMENTED** [review]
I agree it's best remove this conversion.

I've [commented with reasoning in the issue](https://github.com/apache/airflow/issues/15016#issuecomment-837110088) and notes on how I think the logic came to be this way.

But I'm a little unsure about how we should release, given that it is a breaking change, howevery narrowly.  I think this requeires that we create a new major version of the odbc provider.

And it think we _could_ optionally first fix the `if` statement by adding an `else return val` in the current provider major version, before  removing `clean_bool` in next.  @turbaszek or @eladkal do you have advice here?  I'm not sure of the intricacies of bumping provider versions but would be happy to learn so that I can help out with this kind of thing.

-----

Some detail on the way in which this is "breaking":

From pyodbc.connect docstring:

```
      autocommit
        If False or zero, the default, transactions are created automatically as
        defined in the DB API 2.  If True or non-zero, the connection is put into
        ODBC autocommit mode and statements are committed automatically.
```

If someone is using `connect_kwargs` to set `autocommit` using the string value `'false'` then, before this change, autocommit would be `False`.  But after this change, it would be left as string `'False'`, and therefore it would be non-zero and non-False, and therefore the connection would be created with autocommit _on_.

On the other hand, users would **not** be affected if instead they managed autocommit by either (1) using DBApi methods, or (2) explicitly modifying the connection object, or (3) they used actual boolean values instead of strings.  And FWIW this hook has not been released very long.

In any case, I think **this PR needs a line in updating.md**, stating that this hook will no longer convert strings to boolean in `connect_kwargs`.

After that it's good to merge from my perspective, with the caveat that we need a little guidance from others on whether this needs to wait for next version.

**Comment by eladkal** [comment]
> And it think we _could_ optionally first fix the `if` statement by adding an `else return val` in the current provider major >version, before removing `clean_bool` in next

I agree. If we can preserve backward compatibility with deprecation notice we should. we can remove `clean_bool` in the in a followup major release (whenever it may be)

In any case the steps to bump the major version of provider are:

1. bump [provider yaml](https://github.com/apache/airflow/blob/master/airflow/providers/odbc/provider.yaml#L24)
2. add detailed explanation about the change in the [provider changelog](https://github.com/apache/airflow/blob/master/airflow/providers/odbc/CHANGELOG.rst)
@potiuk did I miss something?

**Comment by eladkal** [comment]
@Goodkat are you still working on it?

**Comment by Goodkat** [comment]
@eladkal the implementetion itself is ready to be merged. But the open question is how we should release it.
According to @dstandish recommendation this PR needs a line in updating.md and you also  mentioned provider.yaml/provider changelog.
But I am not sure I should change them myself. Maybe we shall ask someone from the release-preparation team?

**Comment by potiuk** [comment]
> But I am not sure I should change them myself. Maybe we shall ask someone from the release-preparation team?

That would be me :) 

I am about to clarify the rules there (I am **just** finalizing automation around that in #16405  and #16419 ). They were not fully hashed-out before because we were learning about all the different scenarios there, and I think we have observed pretty much all of them so we are ready to codify that for providers.

For a provider, as an author you only need to add description in provider' CHANGELOG.rst when you have breaking change and want to provide a longer description on how people should migrate. There is no need to bump `provider.yaml` (this should be done by release manager based on what kind of changes are prepared). It's actually harmful to do it - we will (if we switch to semantic/conventional commits) will be able to bump those automatically during release preparation for all providers. For now it's  release manager job (and we have tools to help with that).

Coming back to the original question - what do we need now. If I understand correctly, we want to add breaking change where those params have to be 'real' bools not strings. We are releasing breaking changes anyway now for all providers (due to removal of apply_defaults) so adding this now is the best time. I will merge it now, and if you can confirm @Goodkat that this is what the users should do, I will add appropriate description in CHANGELOG.

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
BTW. The main reason why there were strings before, I thing was because this is generally how most other extras are passed anyway. And in case we had it defined in the UI (separate UI field for autocommit and ansi), we would have no way to specifying type of those fields (they are always strings). But it would not be possible for ODBC Hook anyway because it uses dict structure for "connect_kwargs" and this is not supported at all. There is quite a bit of inconsistency between different ways extras are handled by different connections and I think in 3.0 we will have to redefine it a bit.

**Comment by Goodkat** [comment]
@potiuk thank you for the detailed clarification. I confirm that the basic idea was to amend the parameters' types by means of using 'real' bools not strings.

**Comment by potiuk** [comment]
My changelog proposal:

When you pass kwargs to the connection (for example `autocommit` and `ansi`) in `connect_kwargs`
extra you should bass those as booleans. Previously strings were also supported.

```
   "connect_kwargs": {
      "autocommit": "false",
      "ansi": "true"
   }
```

should become

```
   "connect_kwargs": {
      "autocommit": False,
      "ansi": True
   }
```

Does it look good @Goodkat ?

**Comment by Goodkat** [comment]
@potiuk I would better use true/false without capital letters:

'''json
   "connect_kwargs": {
      "autocommit": false,
      "ansi": true
   }
'''

**Comment by potiuk** [comment]
aaaaaaaa Python, python everywhere :).

**Comment by potiuk** [comment]
Corrected. Thanks!

**Pull request merged**
